File: extension/src/engine/scoring.ts
import { CodeBlock, RiskSignal, SignalType } from "./types";
import { matchesAny } from "./glob";

export function isSensitive(fsPath: string, sensitivePaths: string[]): boolean {
  return matchesAny(fsPath, sensitivePaths);
}

/**
 * Compute the review-risk for a single block.
 * Returns null when the block carries no risk (typed, or already reviewed).
 *
 * Risk = external-origin AND unreviewed, weighted by size and sensitivity.
 * This is the TS port of `scoring.rs`.
 */
export function scoreBlock(block: CodeBlock, sensitive: boolean): RiskSignal | null {
  const external = block.origin === "pasted" || block.origin === "ai";
  if (!external) {
    return null;
  }
  // Only an explicit "Mark as reviewed" action clears a block.
  if (block.read) {
    return null;
  }

  const lines = block.endLine - block.startLine + 1;
  // Size drives a 0.5–0.8 base; sensitivity pushes it into critical territory.
  let score = 0.5 + Math.min(lines / 40, 1) * 0.3;
  if (sensitive) {
    score = Math.min(1, score + 0.25);
  }

  const aiLike = block.origin === "ai";
  const originLabel = aiLike ? "AI-generated" : "pasted";

  let type: SignalType;
  if (sensitive) {
    type = aiLike ? "UNREAD_AI_IN_SENSITIVE" : "UNREAD_PASTE_IN_SENSITIVE";
  } else {
    type = aiLike ? "LARGE_UNREAD_AI_BLOCK" : "BULK_PASTE_UNREAD";
  }

  const where = sensitive ? " in a security-sensitive file" : "";
  return {
    type,
    score,
    startLine: block.startLine,
    endLine: block.endLine,
    origin: block.origin,
    message: `${lines} line${lines === 1 ? "" : "s"} of ${originLabel} code${where} not yet reviewed.`,
    suggestion: "Read through this block (or edit/run it) before committing.",
  };
}

File: engine/src/scoring.rs
use crate::provenance::attention::is_reviewed;
use crate::types::{CodeBlock, Origin, RiskSignal, SignalType};

/// Compute the review-risk for a single block, or `None` if it carries no risk
/// (typed origin, or already reviewed). Port of the TS `scoring.ts`.
pub fn score_block(block: &CodeBlock, sensitive: bool) -> Option<RiskSignal> {
    if !block.origin.is_external() || is_reviewed(block) {
        return None;
    }

    let lines = block.end_line.saturating_sub(block.start_line) + 1;
    let mut score = 0.5 + (lines as f32 / 40.0).min(1.0) * 0.3;
    if sensitive {
        score = (score + 0.25).min(1.0);
    }

    let ai_like = block.origin == Origin::Ai;
    let origin_label = if ai_like { "AI-generated" } else { "pasted" };

    let signal_type = if sensitive {
        if ai_like {
            SignalType::UnreadAiInSensitive
        } else {
            SignalType::UnreadPasteInSensitive
        }
    } else if ai_like {
        SignalType::LargeUnreadAiBlock
    } else {
        SignalType::BulkPasteUnread
    };

    let where_ = if sensitive {
        " in a security-sensitive file"
    } else {
        ""
    };
    let plural = if lines == 1 { "" } else { "s" };

    Some(RiskSignal {
        signal_type,
        score,
        start_line: block.start_line,
        end_line: block.end_line,
        origin: block.origin,
        message: format!("{lines} line{plural} of {origin_label} code{where_} not yet reviewed."),
        suggestion: "Read through this block (or edit/run it) before committing.".to_string(),
    })
}
